src/clusterfuzz/_internal/base/task_rate_limiting.py

```diff
@@ -0,0 +1,95 @@
+# Copyright 2024 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Task rate limiting."""
+
+import datetime
+
+from clusterfuzz._internal.datastore import data_types
+from clusterfuzz._internal.datastore import ndb_utils
+from clusterfuzz._internal.metrics import logs
+from clusterfuzz._internal.system import environment
+
+
+def _get_datetime_now():
+  return datetime.datetime.now()
+
+
+class TaskRateLimiter:
+  """Rate limiter for tasks."""
+  TASK_RATE_LIMIT_WINDOW = datetime.timedelta(hours=6)
+  TASK_RATE_LIMIT_MAX_ERRORS = 100
+  TASK_RATE_LIMIT_MAX_COMPLETIONS = 2000
+
+  def __init__(self, task_name, task_argument, job_name):
+    self.task_name = task_name
+    self.task_argument = task_argument
+    self.job_name = job_name
+
+  @property
+  def readable_name(self):
+    return ' '.join([self.task_name, self.task_argument, self.job_name])
+
+  def record_task(self, success: bool) -> None:
+    """Records a task and whether it completed succesfully."""
+    if self.task_name in {'uworker_main', 'postprocess', 'preprocess'}:
```

Comment:
can you put these into a constant somewhere? 

they're also duplicated on line 61.